### DISCUSSION TITLE
RequestContextSettings CachePath set, however cache folder is empty

### DISCUSSION THREAD

**Discussion by assopri** [cefsharp/CefSharp#3575] [Q&A]
I'm using Winforms Cefsharp browser. I need to store cache so that I could create browser, login to the website, surfe, dispose browser and finally recreate browser with the same cache/cookies (so don't need to login again). I also need to have isolated cookies/cache for each particular browser instance. So I use the following code:

`  BrowserSettings browserSetting new BrowserSettings ();
            Browser = new ChromiumWebBrowser();
            (Browser as ChromiumWebBrowser).BrowserSettings = browserSetting;
            RequestContextSettings requestContextSettings = new RequestContextSettings();
            requestContextSettings.PersistSessionCookies = true;
            requestContextSettings.PersistUserPreferences = true;
            requestContextSettings.CachePath = @"C:\Users\User\Documents\temp";
            
            (Browser as ChromiumWebBrowser).RequestContext = new RequestContext(requestContextSettings);`
However the cache folder remains empty. By the way, if I use Cef.initialize (so I`m not having the islolated cache), in this case cache is successfully saved to the specified folder. What am I doing wrong?

**Comment by amaitland** [comment]
> When using a RequestContext you must ensure that RequestContextSettings.CachePath is equal to or a child of CefSettings.RootCachePath See  https://github.com/cefsharp/CefSharp/issues/3111#issuecomment-629713608 for details

As per https://github.com/cefsharp/CefSharp/releases/tag/v90.6.50

**Comment by assopri** [comment]
I fixed the code, so now it looks like this, and lot of files appear in cachefolder, however authorization session is not maintained between browser disposal and further reinitialization. What I miss?

```
Browser = new ChromiumWebBrowser();
            (Browser as ChromiumWebBrowser).BrowserSettings = browserSetting;
            //Using RequestContextSettings allows having isolated cache/cookies etc for different ChromiumWebBrowser instances
            RequestContextSettings requestContextSettings = new RequestContextSettings();
            requestContextSettings.PersistSessionCookies = true;
            requestContextSettings.PersistUserPreferences = true;

            CefSharp.Cef.Initialize(new CefSettings() { RootCachePath = @"C:\Users\User\Documents" });

            requestContextSettings.CachePath = @"C:\Users\User\Documents\temp";
            
            (Browser as ChromiumWebBrowser).RequestContext = new RequestContext(requestContextSettings);
```

**Reply by amaitland** [reply]
You **must** call `Cef.Initialzie` before you create your first `ChromiumWebBrowser` instance.

**Reply by assopri** [reply]
Thank you! It started working, however when I call Cef.Initialize one more time (for example if I restart my app) the session is still lost. I thought cache folder keeps everything needed for session restore.

**Reply by amaitland** [reply]
> however when I call Cef.Initialize one more time (for example if I restart my app) the session is still lost. I thought cache folder keeps everything needed for session restore.

The cache will store cookies, local storage, cached files etc.

It's possible that Cef.Shutdown isn't being called correctly. Are there any errors in the log?

https://github.com/cefsharp/CefSharp/wiki/Trouble-Shooting#log-file
